Add text statistics next to the cursor position

xgouchet opened the pull request:
This adds a word count, alpha numeric character count and all character
count in the bottom right corner, with the cursor line and column
position

Fixes #164

mitya57 commented:
Sorry for the delay with reviewing your requests.
The idea looks fine, but I would like to make this be:

* separate from the cursor position panel (maybe you can make this have a different background color?);
* translatable (wrapping this in a `tr()` and adding a translator comment would be nice).

Also, did you perform any performance checks? How slow is it on large (or huge) documents?

I can look at these issues myself, but in this case you will need to wait a bit more.

xgouchet commented:
Hi, good idea about the translation, I'm going to add it. 

I thought about the performance issue, and might add an option to enable / disable this feature. 
I also thought of using some heuristic to guess the wordcount without having to actually parse the whole document (the QEditText already gives the character count, and we can divide this number by an average character per count value).

mitya57 commented on ReText/editor.py:
```diff
@@ -408,15 +438,72 @@ def leaveEvent(self, event):
 		pos = self.mapToParent(event.pos())
 		pos.setX(pos.x() - self.editor.lineNumberAreaWidth())
 		newEvent = QMouseEvent(event.type(), pos,
-		                       event.button(), event.buttons(),
```
This file uses tabs for indentation, but spaces for alignment. Please do not change this.

xgouchet replied:
:heavy_check_mark:

mitya57 commented on ReText/editor.py:
```diff
@@ -334,6 +342,28 @@ def installFakeVimHandler(self):
+				wordCount += 1
+			wasWordCharacter = isWordCharacter
+		end = time.clock()
+		value = int((end - start) * 1000)
```
Are these start / end / value lines a leftover debug? Can you please share your time measurements?

xgouchet replied:
Yes they were left over, sorry. 
On my laptop, for a 100k document (100'000 characters, 3500 lines), I get an average of 40ms to compute the stats

mitya57 commented on ReText/editor.py:
```diff
@@ -408,15 +438,72 @@ def leaveEvent(self, event):
+	def getText(self):
+		if not globalSettings.documentStatsEnabled:
+			return
+		template = '%d w : %d a : %d c'
```
Please make this translatable.

xgouchet replied:
:heavy_check_mark:

xgouchet commented:
I updated the PR with your recommendation : a base class for additional information (LineInfo / Text Info), and updated the text analysis to correct the word count

mitya57 commented:
Thank you!
